components/proxy/src/test/java/com/hotels/styx/startup/ProxyServerSetUpTest.java

```diff
@@ -0,0 +1,257 @@
+import com.hotels.styx.server.HttpServer;
+import com.hotels.styx.server.HttpsConnectorConfig;
+import com.hotels.styx.support.matchers.LoggingTestSupport;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+
+import static com.hotels.styx.api.support.HostAndPorts.freePort;
+import static com.hotels.styx.proxy.plugin.NamedPlugin.namedPlugin;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.IntStream.range;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.fail;
+
+public class ProxyServerSetUpTest {
+    private final int httpPort = freePort();
+    private final int httpsPort = freePort();
+
+    private final ProxyServerConfig proxyServerConfig = new ProxyServerConfig.Builder()
+            .setHttpConnector(new HttpConnectorConfig(httpPort))
+            .setHttpsConnector(new HttpsConnectorConfig.Builder()
+                    .port(httpsPort)
+                    .build())
+            .build();
+
+    private final Configuration config = new MapBackedConfiguration()
+            .set("proxy", proxyServerConfig);
+
+    private HttpServer server;
+    private LoggingTestSupport log;
+
+    @BeforeMethod
+    public void startRecordingLogs() {
+        log = new LoggingTestSupport(ProxyServerSetUp.class);
+    }
+
+    @AfterMethod
+    public void stopRecordingLogs() {
+        log.stop();
+    }
+
+    @AfterMethod
+    public void stopServer() throws TimeoutException {
+        if (server != null) {
+            try {
+                server.stopAsync().awaitTerminated(3, SECONDS);
+            } catch (IllegalStateException e) {
+                // Caused if the server is already in the 'FAILED' state.
+            }
+        }
+    }
+
+    @Test
+    public void createsServer() throws TimeoutException {
+        StyxServerComponents components = new StyxServerComponents.Builder()
+                .styxConfig(new StyxConfig(config))
+                .build();
+
+        PipelineFactory pipelineFactory = mock(PipelineFactory.class);
+        HttpHandler2 pipeline = mock(HttpHandler2.class);
+        when(pipelineFactory.create(components)).thenReturn(pipeline);
+
+        server = new ProxyServerSetUp(pipelineFactory).createProxyServer(components);
+
+        server.startAsync().awaitRunning(3, SECONDS);
+
+        assertThat(server.httpAddress().getPort(), is(httpPort));
+        assertThat(server.httpsAddress().getPort(), is(httpsPort));
```

Comment:
Hmm... we really should be able to run most of the tests without actually having to bind to the TCP ports. Could you have a quick look if the HTTP server connectors could be somehow mocked for testing?